src/ssh/test-sshbridge.c

```diff
@@ -0,0 +1,1358 @@
+  if (!command)
+    command = fixture_cat.ssh_command;
+  env = g_environ_setenv (env, "COCKPIT_SSH_BRIDGE_COMMAND", command, TRUE);
+
+  if (fix && fix->allow_unknown)
+    {
+      env = g_environ_setenv (env, "COCKPIT_SSH_ALLOW_UNKNOWN",
+                              "true", TRUE);
+    }
+
+  knownhosts_file = fix ? fix->knownhosts_file : NULL;
+  if (!knownhosts_file)
+      knownhosts_file = SRCDIR "/src/ssh/mock_known_hosts";
+
+  env = g_environ_setenv (env, "COCKPIT_SSH_KNOWN_HOSTS_FILE",
+                          knownhosts_file, TRUE);
+  return env;
+}
+
+static CockpitTransport *
+start_bridge (gchar **env,
+              gchar **argv)
+{
+  GError *error = NULL;
+  int fds[2];
+
+  if (socketpair (PF_LOCAL, SOCK_STREAM, 0, fds) < 0)
+    g_assert_not_reached ();
+
+  g_spawn_async_with_pipes (BUILDDIR, argv, env,
+                            G_SPAWN_DO_NOT_REAP_CHILD | G_SPAWN_SEARCH_PATH,
+                            spawn_setup, GINT_TO_POINTER (fds[0]),
+                            NULL, NULL, NULL, NULL, &error);
+  g_assert_no_error (error);
+  close (fds[0]);
+
+  return cockpit_pipe_transport_new_fds ("test-ssh", fds[1], fds[1]);
+}
+
+static void
+on_closed_set_flag (CockpitTransport *transport,
+                    const gchar *problem,
+                    gpointer user_data)
+{
+  gboolean *flag = user_data;
+  g_assert_cmpstr (problem, ==, NULL);
+  g_assert (*flag == FALSE);
+  *flag = TRUE;
+}
+
+static void
+setup (TestCase *tc,
+       gconstpointer data)
+{
+  const TestFixture *fixture = data;
+  const gchar *argv[] = { BUILDDIR "/cockpit-ssh", NULL, NULL };
+  gchar **env = NULL;
+  gchar *host = NULL;
+  gchar *knownhosts_data = NULL;
+
+  alarm (TIMEOUT);
+
+  g_assert (fixture != NULL);
+
+  env = setup_env (fixture);
+  setup_mock_sshd (tc, data);
+
+  if (tc->ssh_port)
+    host = g_strdup_printf ("127.0.0.1:%d", tc->ssh_port);
+  else
+    host = g_strdup ("127.0.0.1");
+  argv[1] = host;
+
+  if (fixture && fixture->knownhosts_data)
+    {
+      if (fixture->knownhosts_data[0] != '\0' &&
+          fixture->knownhosts_data[0] != '*' &&
+          !g_str_equal (fixture->knownhosts_data, "authorize"))
+        {
+          knownhosts_data = g_strdup_printf ("[127.0.0.1]:%d %s",
+                                             tc->ssh_port ? (int)tc->ssh_port : 22,
+                                             fixture->knownhosts_data);
+        }
+      else
+        {
+          knownhosts_data = g_strdup (fixture->knownhosts_data);
+        }
+
+      env = g_environ_setenv (env, "COCKPIT_SSH_KNOWN_HOSTS_DATA",
+                              knownhosts_data, TRUE);
+    }
+
+  tc->transport = start_bridge (env, (gchar **) argv);
+  g_signal_connect (tc->transport, "closed", G_CALLBACK (on_closed_set_flag), &tc->closed);
+  g_strfreev (env);
+  g_free (host);
+  g_free (knownhosts_data);
+}
+
+static void
+teardown (TestCase *tc,
+          gconstpointer data)
+{
+  alarm (0);
```

Comment:
might be more robust to disarm `alarm()` after the `WAIT_UNTIL` and the `waitpid()` below?